sklearn/tree/_splitter.pyx

```diff
@@ -817,26 +859,67 @@ cdef inline int node_split_random(
         current_proxy_improvement = criterion.proxy_impurity_improvement()
 
         if current_proxy_improvement > best_proxy_improvement:
+            current_split.n_missing = n_missing
+
+            # if there are no missing values in the training data, during
+            # test time, we send missing values to the branch that contains
+            # the most samples during training time.
+            if has_missing:
+                current_split.missing_go_to_left = missing_go_to_left
+            else:
+                current_split.missing_go_to_left = n_left > n_right
+
             best_proxy_improvement = current_proxy_improvement
             best_split = current_split  # copy
 
+        # Evaluate when there are missing values and all missing values go
+        # to the right node and non-missing values go to the left node.
+        if has_missing:
```

Comment:
This feels like we are mixing in random and best splitting. I.E. this logic is overriding the randomly selected threshold from `rand_uniform`.

Ideally, it'll be nice to have "split on missing values" be randomly selected, but can not see a distribution to sample from that makes sense. Do you see an alternative?

Reply:
Yeah this is one of the issues I briefly wrote in the PR description and related to #28772. I copied that section here just for convenience. 

```
Compared to BestSplitter, there can be an expected cost to doing splits on missing-values, as we can either:

1. choose a random threshold and then randomly send missing-values to left/right, OR
2. send all missing-values to left/right and all non-missing values to right/left

The push of missing values down the tree can be done randomly (i.e. first option), OR the second option can actually be evaluated. There is a computational cost to doing so, but more importantly there is an interpretation tradeoff. The tradeoff imo comes from the assumption of the missing-values:

if missing-completely-at-random (MCAR), then option 1 is ideal because one simply should ignore the missing values or impute them
if missing-at-random (MAR), then option 2 is nice because sometimes the missing-ness of the data can be informative.
```

> I.E. this logic is overriding the randomly selected threshold from rand_uniform.

Rn the code in the PR still uses a random threshold from `rand_uniform`, but this block only changes `missing_go_to_left` with an extra evaluation.

> Ideally, it'll be nice to have "split on missing values" be randomly selected, but can not see a distribution to sample from that makes sense. Do you see an alternative?

I had logic originally to randomly send missing-values down the tree. However, it didn't perform as well on tests where missing-ness was "informative" as expected, since it's basically treating the missing-ness as MCAR ("random"), when it is not. Let's call the case where we evaluate whether missing-ness to left/right is better as "informative" and randomly sending down left/right as "random".

In BestSplitter, we already have the implementation of informative missing-value splits. Now, if we do random missing-value splits, there is a question of what distribution to sample from. Should it be `rand_int(0, 2)` to indicate left, or right child at equal probability? I think this is the most sensible option.

**This discussion tho I can see being a larger discussion in scikit-learn, so should I for now implement this "random" missing-value split instead of "informative" missing-value split in this PR?**

Reply:
## A proposed longer-term solution that aligns with the literature on missing-values


I still am of the opinion that handling missing-values is a user-driven assumption and as such we should expose a user-facing argument for how to handle it. Therefore, the most sensible alternative for ML practitioners is having a keyword argument in `RandomForest*` and `ExtraTrees*` called `missingness_assumption \in ('mcar', 'mar'), or ('random', 'informative')`, which then either sends the missing-values at random down the tree (i.e. `missing_go_to_left` is selected at random, or evaluates whether a left/right child is better to send the missing values. Even with "informative" missing-value splits, `ExtraTrees*` orders of magnitude faster because it doesn't greedily select the threshold, which is the most expensive part.

Reply:
cc: @glemaitre and @OmarManzoor. 

Is the discussion in https://github.com/scikit-learn/scikit-learn/pull/27966#discussion_r1665885532 worth refactoring out to a separate GH issue?

And for this PR, should I make the missing-values split completely at random for now?

Reply:
> Rn the code in the PR still uses a random threshold from rand_uniform, but this block only changes

Is that the case? If the missing-value splits is deemed to be better, then the threshold changes to: `current_split.threshold = INFINITY`, which will force all the non-missing values to the left and missing values to the right.

> Should it be rand_int(0, 2) to indicate left, or right child at equal probability? I think this is the most sensible option.

I do not think this works. For missing-value splits, if all the missing values go left or right, the tree will end up to be the same. I think of it as sampling (`min_feature_value`, `max_feature_value`) and `INFINITY`. I do not see how much to weight the `INFINITY`.

> I had logic originally to randomly send missing-values down the tree. However, it didn't perform as well on tests where missing-ness was "informative" as expected,

In the context of "ExtraTreeClassifer", I do not expect it to perform well, since the splitting is random (by design).

What do you think of removing this branch and not consider missing-value splits at all?

BTW I am okay with the current solution as long as we document it.

Reply:
Ah right you are. Sorry I think my head was thinking about this incorrectly.

> Is that the case? If the missing-value splits is deemed to be better, then the threshold changes to: current_split.threshold = INFINITY, which will force all the non-missing values to the left and missing values to the right.

Yes I ran into this issue and there is no nice solution. Because we use `rand_uniform`, so setting all non-missing to left and missing to right would indicate we somehow either picked the `max_feature_value + 1e-6`, or `min_feature_value - 1e-6`. 

> In the context of "ExtraTreeClassifer", I do not expect it to perform well, since the splitting is random (by design). What do you think of removing this branch and not consider missing-value splits at all? BTW I am okay with the current solution as long as we document it.

I think either i) keeping the behavior of what we currently do as it's just an extra check per randomly selected feature, or ii) removing this if/else branch works. 

I think it does make sense wrt `main` to remove this if/else branch as to keep the computational complexity the same, but I do think it's a rather cheap check and could make sense in the grand scheme of how sklearn handles missing values.

Reply:
See: [d80b60f](https://github.com/scikit-learn/scikit-learn/pull/27966/commits/d80b60f4e47ed8403ebee7d23bbca2382b342d88) where I removed the if/else branching

and [ac6b25a](https://github.com/scikit-learn/scikit-learn/pull/27966/commits/ac6b25abd07d3ddf1b12bdb504472e3839d9784e) where I document the behavior of missing-values in the ExtraTrees setting.